# KPI 1 - Grid-to-Finish Delta - Data Validation and Sanity Checks

In [5]:
import pandas as pd

# read csv file
df_results = pd.read_csv('/Users/frankdong/Documents/Analytics Local/williams-racing-strategies/processed_data/grid-to-finish.csv')

# dataframe basic info
print(df_results.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 301 entries, 0 to 300
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   race_id          301 non-null    int64 
 1   gp_year          301 non-null    int64 
 2   gp_name          301 non-null    object
 3   gp_round         301 non-null    int64 
 4   driver_name      301 non-null    object
 5   constructor      301 non-null    object
 6   constructor_ref  301 non-null    object
 7   is_williams      301 non-null    bool  
 8   start_position   301 non-null    int64 
 9   final_position   301 non-null    int64 
 10  grid_delta       301 non-null    int64 
dtypes: bool(1), int64(6), object(4)
memory usage: 23.9+ KB
None


## Summary of processed dataset 'grid-to-finish.csv'

- Filepath: /Users/frankdong/Documents/Analytics Local/williams-racing-strategies/processed_data/grid-to-finish.csv *(potentially fix from absolute to relative path later?)*
- Range: 301 entries, 0 to 300.
- Columns: 11
- Data types: int64(6), object(4), bool(1) *(objects are strings)*
- Memory usage: 23.9+ KB 

## Column data types

In [6]:
print(df_results.dtypes)

race_id             int64
gp_year             int64
gp_name            object
gp_round            int64
driver_name        object
constructor        object
constructor_ref    object
is_williams          bool
start_position      int64
final_position      int64
grid_delta          int64
dtype: object


## Missing or null values

In [7]:
df_results.isnull().sum() # No nulls present across the dataset!

race_id            0
gp_year            0
gp_name            0
gp_round           0
driver_name        0
constructor        0
constructor_ref    0
is_williams        0
start_position     0
final_position     0
grid_delta         0
dtype: int64

## Check for duplicates

In [8]:
df_results.duplicated().sum() # no duplicates found

0

## Summary statistics

In [9]:
df_results.describe()

race_id      gp_year    gp_round  start_position  final_position  \
count   301.000000   301.000000  301.000000      301.000000      301.000000   
mean    984.026578  2017.232558   11.644518       11.916944       10.893688   
std      28.081179     1.316208    4.302311        4.625626        3.929630   
min     930.000000  2015.000000    5.000000        2.000000        3.000000   
25%     961.000000  2016.000000    9.000000        8.000000        8.000000   
50%     982.000000  2017.000000   12.000000       12.000000       11.000000   
75%    1005.000000  2018.000000   14.000000       16.000000       14.000000   
max    1029.000000  2019.000000   20.000000       21.000000       20.000000   

       grid_delta  
count  301.000000  
mean     1.023256  
std      3.864728  
min    -17.000000  
25%     -1.000000  
50%      1.000000  
75%      4.000000  
max     11.000000

From this we can roughly tell that, 
- Mean grid to finish delta is a gain of 1.02 places, median a gain of 1 place.
- Standard deviation of delta is 3.86 places
- The most places gained between 2015 and 2019, between Williams and its midfield competitors,
    across the selected ten circuits, is 11 places.
- The most places lost, for the same period, circuits and teams, is 17 places. 

In [10]:
df_results.to_csv('/Users/frankdong/Documents/Analytics Local/williams-racing-strategies/processed_data/grid-to-finish-validated.csv', index=False)

## Validation conclusion
- No null values found
- Column data types are correct
- Grid delta logic is confirmed
- Proceed with feature engineering using the provided CSV data in 'grid-to-finish_validated.csv'